runtime/src/storage/mkvs/sync/proof.rs

```diff
@@ -61,6 +64,112 @@ pub struct Proof {
     pub entries: Vec<Option<RawProofEntry>>,
 }
 
+struct ProofNode {
+    serialized: Vec<u8>,
+    children: Vec<Hash>,
+}
+
+/// A Merkle proof builder.
+pub struct ProofBuilder {
+    root: Hash,
+    included: BTreeMap<Hash, ProofNode>,
+}
+
+impl ProofBuilder {
+    /// Create a new proof builder for the given root hash.
+    pub fn new(root: Hash) -> Self {
+        Self {
+            root,
+            included: BTreeMap::new(),
+        }
+    }
+
+    /// Add a node to the set of included nodes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the node is not clean.
+    pub fn include(&mut self, node: &NodeBox) {
+        if !node.is_clean() {
+            panic!("proof builder: node is not clean");
+        }
+
+        // If node is already included, skip.
+        let nh = node.get_hash();
+        if self.included.contains_key(&nh) {
+            return;
+        }
+        let mut pn = ProofNode {
+            serialized: node
+                .compact_marshal_binary()
+                .expect("marshaling node in proof"),
+            children: vec![],
+        };
+
+        // For internal nodes, also include children.
+        if let NodeBox::Internal(nd) = node {
+            let ch = nd
+                .left
+                .borrow()
+                .node
+                .as_ref()
+                .map(|n| n.borrow().get_hash())
+                .unwrap_or(Hash::empty_hash());
+            pn.children.push(ch);
+
+            let ch = nd
+                .right
+                .borrow()
+                .node
+                .as_ref()
+                .map(|n| n.borrow().get_hash())
+                .unwrap_or(Hash::empty_hash());
+            pn.children.push(ch);
+        }
+
+        self.included.insert(nh, pn);
+    }
+
+    /// Build the proof.
+    pub fn build(&self) -> Proof {
+        let mut proof = Proof {
+            untrusted_root: self.root,
+            entries: vec![],
+        };
+        self._build(&mut proof, self.root);
+
+        proof
+    }
+
+    fn _build(&self, p: &mut Proof, h: Hash) {
+        if h.is_empty() {
+            // Append nil for empty nodes.
+            p.entries.push(None);
+            return;
+        }
+
+        match self.included.get(&h) {
+            None => {
+                // Node is not included in this proof, just add hash of subtree.
+                let mut data: Vec<u8> = h.as_ref().to_vec();
+                data.insert(0, PROOF_ENTRY_HASH);
+                p.entries.push(Some(RawProofEntry(data)));
+            }
+            Some(pn) => {
+                // Pre-order traversal, add visited node.
+                let mut data: Vec<u8> = pn.serialized.clone();
```

Comment:
What about:
```rust
let mut data = Vec::with_capacity(h.len()+1);
data.push(PROOF_ENTRY_HASH);
data.extend_from_slice(h.as_ref());

let mut data = Vec::with_capacity(pn.serialized.len() + 1);
data.push(PROOF_ENTRY_FULL);
data.extend_from_slice(&pn.serialized);
```
this seems clearer to me, and *it might* have even less overhad?

Reply:
Definitely better :+1: